crates/bevy_ecs/src/world/mod.rs

```diff
@@ -86,23 +87,79 @@ use unsafe_world_cell::{UnsafeEntityCell, UnsafeWorldCell};
+///
+/// ## Implementation
+///
+/// The `Marker` generic is necessary to allow multiple blanket implementations
+/// of `Command` for closures, like so (simplified):
+/// ```ignore (This would conflict with the real implementations)
+/// impl Command for FnOnce(&mut World)
+/// impl Command<Result> for FnOnce(&mut World) -> Result
+/// ```
+/// Without the generic, Rust would consider the two implementations to be conflicting.
+///
+/// The type used for `Marker` has no connection to anything else in the implementation.
+pub trait Command<Marker = ()>: Send + 'static {
     /// Applies this command, causing it to mutate the provided `world`.
     ///
     /// This method is used to define what a command "does" when it is ultimately applied.
     /// Because this method takes `self`, you can store data or settings on the type that implements this trait.
     /// This data is set by the system or other source of the command, and then ultimately read in this method.
-    fn apply(self, world: &mut World);
+    fn apply(self, world: &mut World) -> Result;
+
+    /// Returns a new [`Command`] that, when applied, will apply the original command
+    /// and handle any resulting errors according to the provided [`CommandErrorMode`].
+    fn with_error_handling(
+        self,
+        error_mode: CommandErrorMode,
+    ) -> impl FnOnce(&mut World) + Send + 'static
+    where
+        Self: Sized,
+    {
+        move |world: &mut World| {
+            if let Err(error) = self.apply(world) {
+                match error_mode {
+                    CommandErrorMode::Silent => (),
+                    CommandErrorMode::Warning => warn!("Command returned an error: {error}"),
+                    CommandErrorMode::Error => error!("Command returned an error: {error}"),
+                    CommandErrorMode::Panic => panic!("Command returned an error: {error}"),
+                    CommandErrorMode::Custom(function) => function(world, Err(error)),
+                }
+            }
+        }
+    }
+}
+
+/// How a [`Command`] should respond upon encountering an [`Error`](crate::result::Error).
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum CommandErrorMode {
+    /// Upon encountering an error, the command will fail silently.
+    Silent,
+    /// Upon encountering an error, the command will send a warning with [`warn!`].
+    Warning,
+    /// Upon encountering an error, the command will send an error message with [`error!`].
+    Error,
+    /// Upon encountering an error, the command will start a panic.
+    Panic,
+    /// Upon encountering an error, the command will call a custom function,
+    /// passing in its `&mut World` access and the error.
+    ///
+    /// `fn()` can be a closure if it doesn't capture its environment.
+    ///
+    /// If `CommandErrorMode` is used in a situation without `&mut World` access
+    /// (e.g. a non-deferred [`Commands`] method), this mode will instead panic.
+    Custom(fn(&mut World, Result)),
```

Comment:
I thought about this a bit more: this should actually have a `Box<dyn Error>`, not a `Result`. There's no need for error handling if everything went okay, so we can save our users some repetitive matching.

Then, combined with the ideas about how `commands.entity(my_entity)` shouldn't panic, we get:

```rust
enum CommandError {
    NoSuchEntity(Entity),
    CommandFailed(Box<dyn Error>)
}
```

which we pass to the custom error handlers to give them the context needed to solve the problem.